metals/src/main/scala/scala/meta/internal/bsp/BspConnector.scala

```diff
@@ -119,84 +136,183 @@ class BspConnector(
     recursive(root, List.empty)
   }
 
+  /**
+   * Have the user choose what server they'd like to use out of a list of
+   * possible servers that are available to them in the workspace.
+   *
+   * @param possibleServers This could be servers that already have a bsp
+   * entry or just build tools that the user is using that can also be a build
+   * server. This is why we're working with strings instead of
+   * bspConnectionDetails.
+   * @param currentBsp a possible current choice they've made to explicit use
+   * this server in the past.
+   * @return
+   */
   private def askUser(
-      availableBspConnections: List[BspConnectionDetails],
-      currentBsp: Option[String]
-  ): Future[BspResolvedResult] = {
-    val query = Messages.SelectBspServer.request(
-      availableBspConnections,
-      currentBsp
+      possibleServers: List[String],
+      currentSelectedServer: Option[String]
+  ): Future[Option[String]] = {
+    val params = Messages.ChooseBspServer.request(
+      possibleServers,
+      currentSelectedServer
     )
+
     for {
-      item <- client.showMessageRequest(query.params).asScala
-    } yield {
-      val chosenMaybe = Option(item).flatMap(i => query.details.get(i.getTitle))
-      chosenMaybe
-        .map(BspResolvedResult.fromDetails)
-        .getOrElse(ResolvedNone)
-    }
+      item <- client.showMessageRequest(params.params).asScala
+    } yield Option(item).map(_.getTitle()).map(params.mapping(_))
   }
 
   /**
-   * Runs "Switch build server" command, returns true if build server was changed
+   * Runs "Switch build server" command, returns true if build server choice
+   * was changed.
+   *
+   * NOTE: that in most cases this doesn't actaully change your build server
+   * and connect to it, but stores that you want to chage it unless you are
+   * choosing Bloop, since in that case it's special cased and does start it.
    */
   def switchBuildServer(
       workspace: AbsolutePath,
       createBloopAndConnect: () => Future[BuildChange]
   ): Future[Boolean] = {
-    val bloopPresent = buildTools.isBloop
 
-    val availableServers = {
-      val found = bspServers.findAvailableServers()
+    val foundServers = bspServers.findAvailableServers()
+    val bloopPresent: Boolean = buildTools.isBloop
+
+    // These are buildTools in the workspace that can serve as a build servers
+    // and don't already have a .bsp entry
+    val possibleServers: Map[String, Either[
+      BuildTool with BuildServerProvider,
```

Comment:
Also, can't we just do 
```suggestion
      BuildServerProvider,
```
?

Reply:
I thought so too, but then `runUnconditionally` compalins that it has a `BuildServerProvider` and wants a `BuildTool`.

Reply:
Hmm... so maybe we can just do `trait BuildServerProvider extends BuildTool { ` ? Any reason for self type annotation actually?

Reply:
Ah true, I didn't think of that. When I first made it I only included the self type since I modeled it after the `BloopInstallProvider` which also has a self type. So I just followed suit since they are used pretty much exactly the same.